File: backend/app/schemas/founder_ops.py
from datetime import date, datetime, timedelta, timezone
from typing import Any

from pydantic import BaseModel, Field, field_validator, model_validator

from app.schemas.payments import MANUAL_PAYMENT_RAILS
from app.schemas.limits import JsonBounds, StrictInputModel, validate_bounded_json_object


FOUNDER_JSON_BOUNDS = JsonBounds(
    max_container_depth=4,
    max_dict_items=50,
    max_list_items=100,
    max_string_length=1000,
    max_total_bytes=4096,
)


def _validate_founder_json_object(value: dict[str, Any]) -> dict[str, Any]:
    return validate_bounded_json_object(value, bounds=FOUNDER_JSON_BOUNDS)


def _validate_optional_founder_json_object(value: dict[str, Any] | None) -> dict[str, Any] | None:
    if value is None:
        return None
    return _validate_founder_json_object(value)


class AnalyticsEventIn(StrictInputModel):
    event_name: str = Field(min_length=2, max_length=80)
    anonymous_id: str | None = Field(default=None, max_length=120)
    session_id: str | None = Field(default=None, max_length=120)
    subject_id: int | None = Field(default=None, ge=1)
    topic_id: int | None = Field(default=None, ge=1)
    topic_item_id: int | None = Field(default=None, ge=1)
    resource_id: int | None = Field(default=None, ge=1)
    live_session_id: int | None = Field(default=None, ge=1)
    professor_user_id: int | None = Field(default=None, ge=1)
    value_int: int = Field(default=1, ge=0, le=1_000_000)
    duration_seconds: int = Field(default=0, ge=0, le=24 * 60 * 60)
    properties: dict[str, Any] = Field(default_factory=dict)
    occurred_at: datetime | None = None

    @field_validator("event_name", "anonymous_id", "session_id")
    @classmethod
    def normalize_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None

    @field_validator("properties")
    @classmethod
    def validate_properties(cls, value: dict[str, Any]) -> dict[str, Any]:
        return _validate_founder_json_object(value)

    @field_validator("occurred_at")
    @classmethod
    def validate_occurred_at(cls, value: datetime | None) -> datetime | None:
        if value is None:
            return None
        now = datetime.now(timezone.utc)
        occurred_at = value if value.tzinfo is not None else value.replace(tzinfo=timezone.utc)
        if occurred_at > now + timedelta(minutes=10):
            raise ValueError("occurred_at cannot be in the future")
        if occurred_at < now - timedelta(days=370):
            raise ValueError("occurred_at is too old")
        return occurred_at


class AnalyticsEventOut(BaseModel):
    id: int
    event_name: str
    user_id: int | None = None
    occurred_at: datetime
    received_at: datetime


class FinanceExpenseIn(StrictInputModel):
    expense_month: date | None = None
    expense_date: date
    category: str = Field(min_length=2, max_length=60)
    vendor: str | None = Field(default=None, max_length=120)
    description: str | None = Field(default=None, max_length=255)
    amount_centimes: int = Field(ge=0)
    source: str = Field(default="manual", min_length=2, max_length=20)
    status: str = Field(default="paid", min_length=2, max_length=20)
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator("category", "vendor", "description", "source", "status")
    @classmethod
    def normalize_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        return value.strip().lower() if value.strip().lower() in {"manual", "vendor", "estimate", "planned", "paid", "cancelled"} else value.strip()

    @field_validator("source")
    @classmethod
    def validate_source(cls, value: str) -> str:
        if value not in {"manual", "vendor", "estimate"}:
            raise ValueError("source must be manual, vendor, or estimate")
        return value

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        if value not in {"planned", "paid", "cancelled"}:
            raise ValueError("status must be planned, paid, or cancelled")
        return value

    @field_validator("metadata")
    @classmethod
    def validate_metadata(cls, value: dict[str, Any]) -> dict[str, Any]:
        return _validate_founder_json_object(value)


class FinanceExpenseOut(BaseModel):
    id: int
    expense_month: date
    expense_date: date
    category: str
    vendor: str
    description: str
    amount_centimes: int
    currency: str
    source: str
    status: str
    created_by_user_id: int
    metadata: dict[str, Any]
    created_at: datetime
    updated_at: datetime


class RedemptionCodeTemplateIn(StrictInputModel):
    name: str = Field(min_length=2, max_length=160)
    plan: str = Field(default="pro", min_length=1, max_length=60)
    tier: str = Field(default="pro", min_length=1, max_length=30)
    subject_scope: str = Field(default="all", min_length=2, max_length=20)
    subject_ids: list[int] = Field(default_factory=list, max_length=50)
    duration_days: int = Field(default=30, ge=0, le=3700)
    amount_centimes: int = Field(ge=0)
    status: str = Field(default="active", min_length=2, max_length=20)
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator("name", "plan", "tier", "subject_scope", "status")
    @classmethod
    def normalize_text(cls, value: str) -> str:
        return value.strip()

    @field_validator("subject_scope")
    @classmethod
    def validate_subject_scope(cls, value: str) -> str:
        normalized = value.lower()
        if normalized not in {"all", "selected"}:
            raise ValueError("subject_scope must be all or selected")
        return normalized

    @field_validator("status")
    @classmethod
    def validate_template_status(cls, value: str) -> str:
        normalized = value.lower()
        if normalized not in {"active", "archived"}:
            raise ValueError("status must be active or archived")
        return normalized

    @field_validator("metadata")
    @classmethod
    def validate_metadata(cls, value: dict[str, Any]) -> dict[str, Any]:
        return _validate_founder_json_object(value)

    @model_validator(mode="after")
    def validate_subject_selection(self) -> "RedemptionCodeTemplateIn":
        subject_ids = []
        seen: set[int] = set()
        for value in self.subject_ids:
            subject_id = int(value)
            if subject_id <= 0:
                raise ValueError("subject_ids must contain positive integers")
            if subject_id not in seen:
                seen.add(subject_id)
                subject_ids.append(subject_id)
        if self.subject_scope == "selected" and not subject_ids:
            raise ValueError("selected templates require at least one subject")
        self.subject_ids = subject_ids if self.subject_scope == "selected" else []
        return self


class RedemptionCodeTemplateOut(BaseModel):
    id: int
    name: str
    plan: str
    tier: str
    subject_scope: str
    subject_ids: list[int]
    duration_days: int
    amount_centimes: int
    currency: str
    status: str
    created_by_user_id: int
    metadata: dict[str, Any]
    created_at: datetime
    updated_at: datetime


class StaffPaymentProfileOut(BaseModel):
    user_id: int
    display_name: str
    status: str
    monthly_code_limit: int
    monthly_amount_limit_centimes: int
    allowed_template_ids: list[int]
    used_codes_this_month: int
    remaining_codes_this_month: int
    used_amount_this_month_centimes: int
    remaining_amount_this_month_centimes: int | None = None


class StaffPaymentProfileUpdateIn(StrictInputModel):
    display_name: str | None = Field(default=None, max_length=160)
    status: str | None = Field(default=None, min_length=2, max_length=20)
    monthly_code_limit: int | None = Field(default=None, ge=0, le=100_000)
    monthly_amount_limit_centimes: int | None = Field(default=None, ge=0)
    allowed_template_ids: list[int] | None = Field(default=None, max_length=200)
    metadata: dict[str, Any] | None = None

    @field_validator("display_name", "status")
    @classmethod
    def normalize_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.lower()
        if normalized not in {"active", "paused"}:
            raise ValueError("status must be active or paused")
        return normalized

    @field_validator("allowed_template_ids")
    @classmethod
    def normalize_allowed_template_ids(cls, value: list[int] | None) -> list[int] | None:
        if value is None:
            return None
        normalized: list[int] = []
        seen: set[int] = set()
        for raw in value:
            template_id = int(raw)
            if template_id <= 0:
                raise ValueError("allowed_template_ids must contain positive integers")
            if template_id not in seen:
                seen.add(template_id)
                normalized.append(template_id)
        return normalized

    @field_validator("metadata")
    @classmethod
    def validate_metadata(cls, value: dict[str, Any] | None) -> dict[str, Any] | None:
        return _validate_optional_founder_json_object(value)


class StaffPaymentRequestCreateIn(StrictInputModel):
    template_id: int = Field(gt=0)
    payment_method: str = Field(min_length=1, max_length=40)
    provider_reference: str = Field(min_length=3, max_length=160)
    amount_centimes: int = Field(gt=0)
    student_name: str = Field(min_length=2, max_length=160)
    student_phone: str = Field(min_length=4, max_length=80)
    student_email: str | None = Field(default=None, max_length=254)
    proof_url: str | None = Field(default=None, max_length=2000)
    notes: str | None = Field(default=None, max_length=1000)

    @field_validator("payment_method")
    @classmethod
    def normalize_payment_method(cls, value: str) -> str:
        normalized = value.strip().lower().replace("-", "_")
        if normalized not in MANUAL_PAYMENT_RAILS:
            supported = ", ".join(sorted(MANUAL_PAYMENT_RAILS))
            raise ValueError(f"payment_method must be one of: {supported}")
        return normalized

    @field_validator("provider_reference")
    @classmethod
    def normalize_provider_reference(cls, value: str) -> str:
        normalized = "".join(value.strip().upper().split())
        if len(normalized) < 3:
            raise ValueError("provider_reference is required")
        return normalized

    @field_validator("student_name", "student_phone", "student_email", "proof_url", "notes")
    @classmethod
    def normalize_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class RedemptionCodeOut(BaseModel):
    id: int
    code: str
    template_id: int
    generated_by_user_id: int
    redeemed_by_user_id: int | None = None
    plan: str
    tier: str
    subject_ids: list[int]
    duration_days: int
    amount_centimes: int
    currency: str
    status: str
    expires_at: datetime | None = None
    redeemed_at: datetime | None = None
    created_at: datetime


class StaffPaymentRequestOut(BaseModel):
    id: int
    staff_user_id: int
    template_id: int
    redemption_code_id: int
    payment_method: str
    provider_reference: str
    amount_centimes: int
    currency: str
    status: str
    student_name: str
    student_phone: str
    student_email: str
    proof_url: str
    notes: str
    requires_review: bool
    metadata: dict[str, Any]
    created_at: datetime
    updated_at: datetime
    code: RedemptionCodeOut


class StaffPaymentDashboardOut(BaseModel):
    generated_at: datetime
    profile: StaffPaymentProfileOut
    templates: list[RedemptionCodeTemplateOut]
    requests: list[StaffPaymentRequestOut]


class RedemptionCodeRedeemIn(StrictInputModel):
    code: str = Field(min_length=6, max_length=80)

    @field_validator("code")
    @classmethod
    def normalize_code(cls, value: str) -> str:
        normalized = "".join(ch for ch in value.strip().upper() if ch.isalnum())
        if not normalized:
            raise ValueError("code is required")
        return normalized


class RedemptionCodeRedeemOut(BaseModel):
    code: RedemptionCodeOut
    transaction_id: int
    entitlement_count: int


class FounderMetricOut(BaseModel):
    key: str
    label: str
    value: int | float
    previous_value: int | float = 0
    unit: str = "count"


class FounderDashboardOut(BaseModel):
    generated_at: datetime
    month: date
    metrics: list[FounderMetricOut]
    growth_by_day: list[dict[str, Any]]
    students_by_status: dict[str, int]
    students_by_tier: dict[str, int]
    students_by_track: dict[str, int]
    finance: dict[str, Any]
    engagement: dict[str, Any]
    messages: dict[str, Any]
    staff_codes: dict[str, Any]
    expenses: list[FinanceExpenseOut]

File: frontend/lib/founderOps.ts
import { getJson, postJson, putJson } from '@/lib/apiClient'

export type FounderMetric = {
  key: string
  label: string
  value: number
  previous_value: number
  unit: 'count' | 'centimes' | string
}

export type FounderFinanceSummary = {
  paid_revenue_centimes: number
  previous_paid_revenue_centimes: number
  staff_collected_revenue_centimes: number
  staff_redeemed_revenue_centimes: number
  expenses_centimes: number
  open_refunds_centimes: number
  profit_centimes: number
  mrr_centimes: number
  arr_centimes: number
  paid_users: number
  active_entitlements: number
  expenses_by_category: Record<string, number>
  revenue_by_rail: Record<string, number>
  revenue_by_plan: Record<string, number>
}

export type FounderEngagementSummary = {
  active_students_7d: number
  video_events_month: number
  approx_video_watch_minutes: number
  live_sessions_month: number
  live_joined_students_month: number
  live_questions_month: number
  quiz_attempts_month: number
  total_xp: number
  ai_quota_units_month: number
}

export type FounderMessagesSummary = {
  private_conversations: number
  private_messages_month: number
  unread_for_professors: number
  professors_with_chats: number
}

export type FounderStaffCodesSummary = {
  generated_month: number
  redeemed_month: number
  unused_total: number
  collected_staff_revenue_centimes: number
  redeemed_staff_revenue_centimes: number
}

export type FinanceExpense = {
  id: number
  expense_month: string
  expense_date: string
  category: string
  vendor: string
  description: string
  amount_centimes: number
  currency: string
  source: string
  status: string
  created_by_user_id: number
  metadata: Record<string, unknown>
  created_at: string
  updated_at: string
}

export type FinanceExpenseInput = {
  expense_date: string
  expense_month?: string
  category: string
  vendor?: string
  description?: string
  amount_centimes: number
  source?: 'manual' | 'vendor' | 'estimate'
  status?: 'planned' | 'paid' | 'cancelled'
  metadata?: Record<string, unknown>
}

export type RedemptionCodeTemplateInput = {
  name: string
  plan: string
  tier: string
  subject_scope: 'all' | 'selected'
  subject_ids: number[]
  duration_days: number
  amount_centimes: number
  status?: 'active' | 'archived'
  metadata?: Record<string, unknown>
}

export type FounderDashboard = {
  generated_at: string
  month: string
  metrics: FounderMetric[]
  growth_by_day: Array<Record<string, unknown>>
  students_by_status: Record<string, number>
  students_by_tier: Record<string, number>
  students_by_track: Record<string, number>
  finance: FounderFinanceSummary
  engagement: FounderEngagementSummary
  messages: FounderMessagesSummary
  staff_codes: FounderStaffCodesSummary
  expenses: FinanceExpense[]
}

export type RedemptionCodeTemplate = {
  id: number
  name: string
  plan: string
  tier: string
  subject_scope: string
  subject_ids: number[]
  duration_days: number
  amount_centimes: number
  currency: string
  status: string
  created_by_user_id: number
  metadata: Record<string, unknown>
  created_at: string
  updated_at: string
}

export type RedemptionCode = {
  id: number
  code: string
  template_id: number
  generated_by_user_id: number
  redeemed_by_user_id: number | null
  plan: string
  tier: string
  subject_ids: number[]
  duration_days: number
  amount_centimes: number
  currency: string
  status: string
  expires_at: string | null
  redeemed_at: string | null
  created_at: string
}

export type StaffPaymentRequest = {
  id: number
  staff_user_id: number
  template_id: number
  redemption_code_id: number
  payment_method: string
  provider_reference: string
  amount_centimes: number
  currency: string
  status: string
  student_name: string
  student_phone: string
  student_email: string
  proof_url: string
  notes: string
  requires_review: boolean
  metadata: Record<string, unknown>
  created_at: string
  updated_at: string
  code: RedemptionCode
}

export type StaffPaymentProfile = {
  user_id: number
  display_name: string
  status: string
  monthly_code_limit: number
  monthly_amount_limit_centimes: number
  allowed_template_ids: number[]
  used_codes_this_month: number
  remaining_codes_this_month: number
  used_amount_this_month_centimes: number
  remaining_amount_this_month_centimes: number | null
}

export type StaffPaymentProfileInput = {
  display_name?: string
  status?: 'active' | 'paused'
  monthly_code_limit?: number
  monthly_amount_limit_centimes?: number
  allowed_template_ids?: number[]
  metadata?: Record<string, unknown>
}

export type StaffPaymentDashboard = {
  generated_at: string
  profile: StaffPaymentProfile
  templates: RedemptionCodeTemplate[]
  requests: StaffPaymentRequest[]
}

export type StaffPaymentRequestInput = {
  template_id: number
  payment_method: string
  provider_reference: string
  amount_centimes: number
  student_name: string
  student_phone: string
  student_email?: string
  proof_url?: string
  notes?: string
}

export const EMPTY_FOUNDER_FINANCE: FounderFinanceSummary = {
  paid_revenue_centimes: 0,
  previous_paid_revenue_centimes: 0,
  staff_collected_revenue_centimes: 0,
  staff_redeemed_revenue_centimes: 0,
  expenses_centimes: 0,
  open_refunds_centimes: 0,
  profit_centimes: 0,
  mrr_centimes: 0,
  arr_centimes: 0,
  paid_users: 0,
  active_entitlements: 0,
  expenses_by_category: {},
  revenue_by_rail: {},
  revenue_by_plan: {},
}

export const EMPTY_FOUNDER_DASHBOARD: FounderDashboard = {
  generated_at: '',
  month: '',
  metrics: [],
  growth_by_day: [],
  students_by_status: {},
  students_by_tier: {},
  students_by_track: {},
  finance: EMPTY_FOUNDER_FINANCE,
  engagement: {
    active_students_7d: 0,
    video_events_month: 0,
    approx_video_watch_minutes: 0,
    live_sessions_month: 0,
    live_joined_students_month: 0,
    live_questions_month: 0,
    quiz_attempts_month: 0,
    total_xp: 0,
    ai_quota_units_month: 0,
  },
  messages: {
    private_conversations: 0,
    private_messages_month: 0,
    unread_for_professors: 0,
    professors_with_chats: 0,
  },
  staff_codes: {
    generated_month: 0,
    redeemed_month: 0,
    unused_total: 0,
    collected_staff_revenue_centimes: 0,
    redeemed_staff_revenue_centimes: 0,
  },
  expenses: [],
}

export const EMPTY_STAFF_PAYMENT_DASHBOARD: StaffPaymentDashboard = {
  generated_at: '',
  profile: {
    user_id: 0,
    display_name: '',
    status: 'active',
    monthly_code_limit: 0,
    monthly_amount_limit_centimes: 0,
    allowed_template_ids: [],
    used_codes_this_month: 0,
    remaining_codes_this_month: 0,
    used_amount_this_month_centimes: 0,
    remaining_amount_this_month_centimes: null,
  },
  templates: [],
  requests: [],
}

export function founderDashboardPath(month?: string) {
  return month ? `/admin/founder-dashboard?month=${encodeURIComponent(month)}` : '/admin/founder-dashboard'
}

export async function getFounderDashboard(month?: string) {
  return getJson<FounderDashboard>(founderDashboardPath(month))
}

export async function createFinanceExpense(input: FinanceExpenseInput) {
  return postJson<FinanceExpense, FinanceExpenseInput>('/admin/finance/expenses', input)
}

export async function getRedemptionTemplates(includeArchived = false) {
  return getJson<RedemptionCodeTemplate[]>(`/admin/redemption-templates?include_archived=${includeArchived ? 'true' : 'false'}`)
}

export async function createRedemptionTemplate(input: RedemptionCodeTemplateInput) {
  return postJson<RedemptionCodeTemplate, RedemptionCodeTemplateInput>('/admin/redemption-templates', input)
}

export async function getStaffPaymentRequests(limit = 100) {
  return getJson<StaffPaymentRequest[]>(`/admin/staff-payment-requests?limit=${limit}`)
}

export async function getStaffPaymentProfiles(limit = 100) {
  return getJson<StaffPaymentProfile[]>(`/admin/staff-payment-profiles?limit=${limit}`)
}

export async function upsertStaffPaymentProfile(userId: number, input: StaffPaymentProfileInput) {
  return putJson<StaffPaymentProfile, StaffPaymentProfileInput>(`/admin/staff-payment-profiles/${userId}`, input)
}

export async function getStaffPaymentDashboard(limit = 50) {
  return getJson<StaffPaymentDashboard>(`/staff/payments/dashboard?limit=${limit}`)
}

export async function createStaffPaymentRequest(input: StaffPaymentRequestInput) {
  return postJson<StaffPaymentRequest, StaffPaymentRequestInput>('/staff/payments/requests', input)
}

export function numberValue(value: unknown): number {
  if (typeof value === 'number') return Number.isFinite(value) ? value : 0
  return 0
}

export function normalizeFounderGrowthRows(rows: Array<Record<string, unknown>>, fallbackTotal: number) {
  if (!rows.length) return rows

  const normalizedRows = rows.map((row) => ({
    ...row,
    new_students: numberValue(row.new_students),
    total_students: typeof row.total_students === 'number' && Number.isFinite(row.total_students)
      ? row.total_students
      : null,
  }))
  const hasPositiveNewStudents = normalizedRows.some((row) => numberValue(row.new_students) > 0)
  const hasPositiveExplicitTotal = normalizedRows.some((row) => numberValue(row.total_students) > 0)
  const hasAnyExplicitTotal = normalizedRows.some((row) => row.total_students !== null)
  const shouldTrustZeroExplicitTotals = !hasPositiveExplicitTotal && (!hasPositiveNewStudents || numberValue(fallbackTotal) === 0)
  const canUseExplicitTotals = hasAnyExplicitTotal && (hasPositiveExplicitTotal || shouldTrustZeroExplicitTotals)
  const isUsableExplicitTotal = (value: number | null) => (
    value !== null && (numberValue(value) > 0 || shouldTrustZeroExplicitTotals)
  )
  const firstExplicitTotalIndex = canUseExplicitTotals
    ? normalizedRows.findIndex((row) => isUsableExplicitTotal(row.total_students))
    : -1
  const openingTotal = firstExplicitTotalIndex >= 0
    ? Math.max(
      0,
      numberValue(normalizedRows[firstExplicitTotalIndex]?.total_students)
        - normalizedRows
          .slice(0, firstExplicitTotalIndex + 1)
          .reduce((sum, row) => sum + numberValue(row.new_students), 0),
    )
    : Math.max(
      0,
      numberValue(fallbackTotal) - normalizedRows.reduce((sum, row) => sum + numberValue(row.new_students), 0),
    )
  let runningTotal = openingTotal

  return normalizedRows.map((row) => {
    runningTotal += numberValue(row.new_students)
    if (canUseExplicitTotals && isUsableExplicitTotal(row.total_students)) runningTotal = numberValue(row.total_students)
    return {
      ...row,
      total_students: runningTotal,
    }
  })
}

export function formatMoneyCentimes(value: unknown) {
  return `${(numberValue(value) / 100).toLocaleString('en-US', {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  })} MAD`
}

export function formatNumber(value: unknown) {
  return numberValue(value).toLocaleString('en-US')
}

export function moneyInputToCentimes(value: string) {
  const trimmed = value.trim().replace(/\s+/g, '')
  if (!trimmed) return null

  const normalized = normalizeMoneyInput(trimmed)
  if (!normalized) return null
  const amount = Number(normalized)
  if (!Number.isFinite(amount) || amount < 0) return null
  return Math.round(amount * 100)
}

function normalizeMoneyInput(value: string) {
  if (/^\d{1,3}(,\d{3})+(\.\d{1,2})?$/.test(value)) {
    return value.replace(/,/g, '')
  }
  if (/^\d{1,3}(\.\d{3})+(,\d{1,2})?$/.test(value)) {
    return value.replace(/\./g, '').replace(',', '.')
  }
  if (value.includes(',') && value.includes('.')) {
    const lastComma = value.lastIndexOf(',')
    const lastDot = value.lastIndexOf('.')
    return lastComma > lastDot
      ? value.replace(/\./g, '').replace(',', '.')
      : value.replace(/,/g, '')
  }
  if (value.includes(',') && !value.includes('.')) {
    return value.replace(',', '.')
  }
  return value
}

export function recordEntries(record?: Record<string, unknown> | null, limit = 8) {
  return Object.entries(record ?? {})
    .map(([key, value]) => ({ key, value: numberValue(value) }))
    .filter((item) => item.value > 0)
    .sort((a, b) => b.value - a.value)
    .slice(0, limit)
}
